### PR TITLE
Comment enhancement

### PR DESCRIPTION
**Pull request by lamtranb**
Hi Frank,

Below are the new feature for StudentQuiz:

- Make it into a proper 'discussion area'; also allow response-comments to be collapsed from view
- Comment Report feature
- Comment sorting
- Tidy up Navigator buttons display

@HuongNV13 had assigned to me and I fixed changes requested as mention [here](https://github.com/frankkoch/moodle-mod_studentquiz/pull/188)
Please help me to review again.

Many thanks,

### CODE REVIEW

**Comment by dionysius** [comment]
For myself:

Add mustache to CI
In-depth manual backup and restore of comments

**Comment by timhunt** [comment]
You may already know this, but in case you have not hit the problem before:

The failure of the "moodle-plugin-ci grunt" check on the Moodle 3.8 branch is expected if you are trying to release a single version of the plugin that works with all Moodle versions 3.5 -> 3.8. They changed the way Moodle builds JavaScript in 3.8 so the results of running grunt are different. However, files built for Moodle 3.7 will work in Moodle 3.8. Therefore you can safely ignore these fails.

And, there is a trick to tell Travis to ingore these failures. You can see an example of me doing it here: https://github.com/moodleou/moodle-qtype_recordrtc/blob/master/.travis.yml#L68

It is also worth noting that the moodle-plugin-ci mustache check often fails with error message " WARNING: Problem calling HTML validator - please report bug to integration team." even if there is nothing wrong with your template files, which is annoying.

I hope that is helpful.

**Comment by dionysius** [comment]
Thank you @timhunt, they are indeed helpful. I'm aware of the grunt related bug and posted the issue a while ago https://github.com/blackboard-open-source/moodle-plugin-ci/issues/109

**Comment by timhunt** [comment]
Sorry, but I was just testing this, and there is a problem with the upgrade code on the branch.

Commit ea21b9e has important changes in db/upgrade.php, but there is no change in version.php, so that upgrade step does not run. Therefore, if you apply this branch to an exsiting Moodle site, and run the upgrade, and then try to use StudentQuiz, you get errors like

ERROR: column "deleted" does not exist

on line 813 of /mod/studentquiz/classes/question/bank/studentquiz_bank_view.php: call to pgsql_native_moodle_database->get_recordset_sql().

**Comment by timhunt** [comment]
https://github.com/timhunt/moodle-mod_studentquiz/commit/ef6be03c4231a6a3a9a74451a7d58509d74403f4 is a commit on top of Lam's branch which fixes the upgrade. Not sure if that is the best way to do it.

**Comment by lamtranb** [comment]
Hi @timhunt 
Sorry for my mistake when I rebase upstream.
I can make a new commit to fix it and squash into [ea21b9](https://github.com/frankkoch/moodle-mod_studentquiz/commit/ea21b9eb40349f8b3f45ab3c8f33aff9100312ff).
And I will update this commit [035fb05](https://github.com/frankkoch/moodle-mod_studentquiz/pull/190/commits/035fb056e0d1f6a4744569da7eedf846e5a834c5#diff-5bbd3ccd25214f9956eea7e9f714bc08) too
May I do that?
Many thanks.

**Comment by timhunt** [comment]
@lamtranb, yes, please. Fix it however you think is best. Thank you for working on this.

**Comment by lamtranb** [comment]
> @lamtranb, yes, please. Fix it however you think is best. Thank you for working on this.

Hi @timhunt
I have fixed it. Please help me to review it again.
Thanks.

**Comment by timhunt** [comment]
I don't have time to do a full review now. It is more important that @frankkoch and @dionysius are happy with the code than what I think. From the little bit of testing we did today, it seems to work, and work quite nicely.

Thank you for all your hard work on this, Lam. Chúc Mừng Năm Mới.

**Comment by frankkoch** [comment]
This looks great to me, thank you for the marvellous work.

**Comment by timhunt** [comment]
Thank you very much for reviewing this so quickly Frank.

File: db/install.xml [hunk 1]
```diff
@@ -1,19 +1,19 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<XMLDB PATH="mod/studentquiz/db" VERSION="20101203" COMMENT="XMLDB file for Moodle mod/studentquiz"
+<XMLDB PATH="mod/studentquiz/db" VERSION="20200116" COMMENT="XMLDB file for Moodle mod/studentquiz"
```
Comment (dionysius) [inline db/install.xml]: @timhunt wow, I never ever touched that field, that must/should be updated constantly whenever you change the database scheme?

**Comment by timhunt** [comment]
The number if install.xml is not terribly important. It is mostly for info. It gest updated automatically if you use the build in XMLDB editor to edit the file, which you alwasy should.

The bit that is important is that the numbers in version.php match the ones in upgrade.php, since that affects whether the upgrade works. (And when it comes to rebasing branches that involve database changes, it is very easy to slip up and make a mistake.)

**Pull request merged**